# SEDR
## The import form of the SEDR package
import SEDR
## Constructs a neighborhood graph for spatial transcriptomics data, which is essential for capturing spatial relationships between cells.
SEDR.graph_construction(adata, k)
### the detail description of parameters:
1. adata: AnnData object
The annotated data matrix containing spatial transcriptomics data. This should include gene expression data and spatial coordinates.
2. k: int (default: 12)
The number of nearest neighbors to consider when constructing the graph
## Initializes a SEDR object, which is used to perform unsupervised learning on spatial transcriptomics data.
SEDR.Sedr(adata, graph_dict, mode)
### the detail description of parameters:
1. adata: AnnData object
The annotated data matrix containing spatial transcriptomics data.
2. graph_dict: dict
The graph dictionary generated by SEDR.graph_construction().
3. mode: str (default: 'imputation')
The mode of operation. Options include 'imputation' for gene expression imputation and 'embedding' for learning a low-dimensional representation
## Trains the SEDR model with the Deep Embedded Clustering (DEC) module, which helps in learning a more robust representation.
SEDR.Sedr.train_with_dec()
## Trains the SEDR model without the DEC module, which can be useful for certain types of analysis.
SEDR.Sedr.train_without_dec()
## Reconstructs the gene expression data using the trained model, which can be useful for imputation tasks.
SEDR.Sedr.recon()
